File: src/utils/mediaCapabilities.ts
import { isSourceLocal, isSourceRemote, type ResolvedTrack } from '../store/player'

export const AUDIO_EXTENSIONS = ['mp3', 'wav', 'ogg', 'flac', 'm4a', 'aac', 'opus']
export const BROWSER_VIDEO_EXTENSIONS = ['mp4', 'm4v', 'webm']
export const EXTERNAL_VIDEO_EXTENSIONS = ['mkv', 'avi', 'mov', 'flv', 'wmv', 'ts', 'm2ts', 'mpg', 'mpeg', '3gp']
export const VIDEO_EXTENSIONS = [...BROWSER_VIDEO_EXTENSIONS, ...EXTERNAL_VIDEO_EXTENSIONS]
export const MEDIA_EXTENSIONS = [...AUDIO_EXTENSIONS, ...VIDEO_EXTENSIONS]

export function getPathExtension(path: string | null | undefined) {
  return path?.split('.').pop()?.toLowerCase() || ''
}

export function getTrackFilePath(track: ResolvedTrack | null | undefined) {
  if (!track) return null
  if (isSourceLocal(track.source)) return track.source.Local.path
  if (isSourceRemote(track.source) && track.source.Remote.cached_path) return track.source.Remote.cached_path
  return null
}

export function trackHasVideo(track: ResolvedTrack | null | undefined) {
  if (!track) return false
  if (track.media_type === 'Video') return true
  if (isSourceRemote(track.source)) return track.source.Remote.media_type === 'Video'
  return VIDEO_EXTENSIONS.includes(getPathExtension(getTrackFilePath(track)))
}

File: src-tauri/src/services/media_capabilities.rs
use crate::models::playlist::MediaType;
use std::path::Path;

pub const AUDIO_EXTENSIONS: &[&str] = &["mp3", "wav", "ogg", "flac", "m4a", "aac", "opus"];
pub const BROWSER_VIDEO_EXTENSIONS: &[&str] = &["mp4", "m4v", "webm"];
pub const EXTERNAL_VIDEO_EXTENSIONS: &[&str] = &[
    "mkv", "avi", "mov", "flv", "wmv", "ts", "m2ts", "mpg", "mpeg", "3gp",
];
pub const MEDIA_EXTENSIONS: &[&str] = &[
    "mp3", "wav", "ogg", "flac", "m4a", "aac", "opus", "mp4", "m4v", "webm", "mkv",
    "avi", "mov", "flv", "wmv", "ts", "m2ts", "mpg", "mpeg", "3gp",
];

pub fn extension_lower(path: &Path) -> Option<String> {
    path.extension().map(|ext| ext.to_string_lossy().to_lowercase())
}

pub fn is_audio_path(path: &Path) -> bool {
    extension_lower(path)
        .map(|ext| AUDIO_EXTENSIONS.contains(&ext.as_str()))
        .unwrap_or(false)
}

pub fn is_browser_video_path(path: &Path) -> bool {
    extension_lower(path)
        .map(|ext| BROWSER_VIDEO_EXTENSIONS.contains(&ext.as_str()))
        .unwrap_or(false)
}

pub fn is_external_video_path(path: &Path) -> bool {
    extension_lower(path)
        .map(|ext| EXTERNAL_VIDEO_EXTENSIONS.contains(&ext.as_str()))
        .unwrap_or(false)
}

pub fn is_video_path(path: &Path) -> bool {
    is_browser_video_path(path) || is_external_video_path(path)
}

pub fn is_supported_media_path(path: &Path) -> bool {
    is_audio_path(path) || is_video_path(path)
}

pub fn media_type_from_path(path: &Path) -> MediaType {
    if is_video_path(path) {
        MediaType::Video
    } else {
        MediaType::Audio
    }
}
